### PR TITLE
Add plutus-streaming package

### PR DESCRIPTION
**Pull request by andreabedini**
Hi there,

this brings into main the plutus-streaming "library" I had written. I am using quotes because it's effectively one function :D

~It still has one bug where the user gets a STM exception if the chain-sync client cannot connect. I need to find a way to handle this.~ solved.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-streaming/src/Plutus/Streaming.hs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            localNodeNetworkId = networkId,
+            localNodeSocketPath = socketPath
+          }
+
+      -- FIXME this comes from the config file but Cardano.Api does not expose readNetworkConfig!
+      epochSlots = EpochSlots 40
+
+      clientThread = do
+        connectToLocalNode connectInfo localNodeClientProtocols
+        -- the only reason the clien can terminate successfully is if it
```
Comment (sjoerdvisscher) [inline plutus-streaming/src/Plutus/Streaming.hs:72]:
```suggestion
        -- the only reason the client can terminate successfully is if it
```

File: plutus-streaming/src/Plutus/Streaming.hs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+-- and runs the chain-sync mini-protocol.
+--
+-- I am using the term "streaming client" because the only things it does
+-- is to keep sending requests for new blocks.
+--
+-- In particular, this client is fire-and-forget and does not require any
+-- control.
+--
+-- Blocks obtained from the chain-sync mini-protocol are passed to a
+-- consumer through a channel. To understand the MVar-Maybe-Chan dance see
```
Comment (sjoerdvisscher) [inline plutus-streaming/src/Plutus/Streaming.hs:96]: Which dance, which note?

Comment (andreabedini) [reply plutus-streaming/src/Plutus/Streaming.hs:96]: Ouch, leftover note, I should have reviewed all the comments. Apologies.

File: plutus-streaming/src/Plutus/Streaming.hs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  NetworkId ->
+  -- | The point on the chain to start streaming from
+  ChainPoint ->
+  (Stream (Of SimpleChainSyncEvent) IO r -> IO b) ->
+  IO b
+withSimpleChainSyncEventStream socketPath networkId point consumer = do
+  -- The chain-sync client runs in a different thread and it will send us
+  -- block through this channel.
+  chan <- newBroadcastTChanIO
+  readerChannel <- atomically $ dupTChan chan
```
Comment (sjoerdvisscher) [inline plutus-streaming/src/Plutus/Streaming.hs:49]: Can you add a comment here why `dupTChan` is needed here. What happens if we just use `chan` instead?

File: plutus-streaming/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Revision history for plutus-streaming
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply plutus-streaming/CHANGELOG.md]: I have removed the CHANGELOG file.

File: plutus-streaming/src/Plutus/Streaming.hs (hunk 2 above)
Comment (andreabedini) [reply plutus-streaming/src/Plutus/Streaming.hs]: Reworked all the comments.

File: plutus-streaming/src/Plutus/Streaming.hs (hunk 3 above)
Comment (andreabedini) [reply plutus-streaming/src/Plutus/Streaming.hs:56]: I added a note linking `newBroadcastTChan` [documentation](https://hackage.haskell.org/package/stm-2.5.0.2/docs/Control-Concurrent-STM-TChan.html#v:newBroadcastTChan).

**Comment by andreabedini** [comment]
I solved the bug. Basically I was using async wrong, exceptions in the parent thread kill the worker thread, but I wanted the opposite.

**Review by sjoerdvisscher — APPROVED** [review]

**Pull request merged**